### DISCUSSION TITLE
[WIP] Suggestion - add `/cohort` endpoint

### DISCUSSION THREAD

**Discussion by e-t-k** [CBIIT/ccdi-federation-api#77] [API Design Discussion]
Hi all, 
The idea has come up a couple times of building a Cohort or Dataset endpoint and as I work on the Treehouse implementation of the API I've encountered a problem where the best solution I can think of is such an endpoint. Here's my proposal for how this interface might go - can we discuss whether this is something we want to add?

### The problem
Currently, a call to the **/file** endpoint returns a paginated list of files. Each file response contains a list of sample identifiers. However, this list of identifiers is not itself paginated.
This works well for files with one or a few samples. At UCSC Treehouse, however, we share some large files containing the RNA expression of over ten thousand samples. The corresponding file endpoint thus contains an unpaginated list of 10k identifiers.

I feel that it is bad practice to send such a large unpaginated response and that, just as we paginate our main file/sample/subject endpoints, we should paginate this list.
One option would be to paginate within the returned file entity -- but I would like to avoid the the design complexities of nested pagination.
Therefore, I'd like to propose an alternative solution that may also be useful for other purposes: a **Cohort** entity.

### Cohort
Cohort: a group of samples associated with a file or files and optionally given a name. This is a deliberately minimalist definition that does not have any baggage (organizations, study names, etc) associated with it.

### Questions to consider 
As you read the rest of this, here are some things I'm not certain about.
Please also suggest any other changes that come to mind or concerns that you have; I'm certain that there are things I failed to consider here.

0. Is a list of 10k namespace-delimited identifiers actually a large response? Is this actually a problem we need to solve?
0. I have designed Cohort as a lightweight sample group. *Is that what we want right now* ? Is there value in going straight to a more formal grouping that reflects a scientific study or more meaningful group of samples?
1. Naming - Possible names could be Cohort, Dataset, something else (sampleList?) I want the name to reflect the informality of the grouping.
2. I feel that a file should have exactly 1 cohort associated with it. Should a file have a list of cohorts instead?
3. Is there a compelling reason to add a `/cohort` endpoint that lists all cohorts known to the server? If so, how would we paginate that?
4. Should cohorts have a simple string for a name or should the name be qualified under `organization/namespace` ?

### Interface
The Cohort entity could appear in the "samples" value of a File entity. Similar to the existing Gateway entity, a "samples" value would contain either an anonymous cohort or a reference to a named cohort.  An anonymous cohort would be almost identical to the current interface, containing a sample list inside the File entity. A named reference, however, would contain no sample list but only a cohort name; to retrieve the sample list, a second API call would be made to a paginated cohort endpoint.

### New endpoint `/cohort/{name}`
Similar to Gateways, I do **not** suggest having a standalone `cohort` endpoint which would list cohorts known to the server. Named cohorts should instead be discoverable via the files which mention them and then be individually queryable once the name is known. In addition, a cohort list endpoint would have the same problem as the current file endpoint of potentially containing long unpaginated data.

Instead, I propose just one new endpoint, for describing an individual cohort: `/cohort/{name}`. This endpoint would return a single Named cohort and its **paginated** list of samples. Pagination would be mandatory and function similarly to existing endpoints. Each page would include the cohort name and summary along with the current page of sample identifier items.

#### Proposed schema

/cohort/{name}

```
200 Successful operation - returns a models.cohort.Named.
404 Not Found - returns a NotFound error. eg:

{
  "errors": [
    {
      "kind": "NotFound",
      "entity": "Cohort with name 'foo'",
      "message": "Cohort with names 'foo' not found."
    }
  ]
}

```

models.cohort.Named {
```
  description: A named Cohort. Named cohorts exist so that a long list of sample identifiers,
    as contained in a File, can be paged through. Sample identifiers must be sorted [same as other endpoints]...

  name*: The name of the cohort.
  summary*: responses.entity.Summary : same as other summaries
  samples* : [ 
    Similar to the 'data' key of other responses, this contains the paginated sample list of the cohort.
    models.sample.Identifier > {...} ]
}
```

models.cohort.AnonymousOrReference {
```
  description: An anonymous list of samples or a reference to a named [Cohort].
  oneOf -> 
    {
      description: An anonymous list of samples.
      samples* : [
        "One or more samples that are associated with this [File]..."
        models.sample.Identifier > {} 
      ]
      kind* : string : Enum [ Anonymous ] 
    }

    {
      description: A reference to a named Cohort.
      samples* : string : The name of the Cohort which lists this file's samples' identifiers.
      kind* : string : Enum [ Reference ] 
    }
```

  

#### Example: an entire named cohort with few samples

```
/cohort/my-example-cohort

{
  "name": "my-example-cohort",
  "summary": {
    "counts": {
      "current": 2,
      "all": 2
    }
  },
  "samples":[
    {
      "namespace": {
        "organization": "example-organization",
          "name": "ExampleNamespaceOne"
      },
      "name": "Sample78"
    },
    {
      "namespace": {
        "organization": "example-organization",
          "name": "ExampleNamespaceOne"
      },
      "name": "Sample79"
    }
  ]
}
```

#### Example: a page from a named cohort with many samples

```
/cohort/example-two?page=3000&per_page=1

{
  "name": "my-example-cohort",
  "summary": {
    "counts": {
      "current": 1,
      "all": 10000
    }
  },
  "samples":[
    {
      "namespace": {
        "organization": "example-organization",
          "name": "ExampleNamespaceOne"
      },
      "name": "Sample3000"
    }
  ]
}
```

### Changed model: File

File responses with anonymous sampleLists (ie, the current default) would be slightly changed. Instead of the `samples` key having a direct list as a value, there would be an extra dict wrapping that list in another `samples` key, and there would be an additional `kind` key so that users could identify the anonymous list.

#### Examples 
File response with anonymous Cohort:

      {
        "id": {
          "namespace": {
            "organization": "example-organization",
            "name": "ExampleNamespaceOne"
          },
          "name": "File102.txt"
        },
        "samples": {
          "kind": "Anonymous",
          "samples": [
            { 
              "namespace": {
                "organization": "example-organization",
                "name": "ExampleNamespaceOne"
              },
              "name": "Sample91"
            }
          ]
        },
        "gateways": [
		# omitted
        ],
        "metadata": {
		# omitted
          }
        }
      }

A file response with a reference to a named Cohort:

      {
        "id": {
          "namespace": {
            "organization": "example-organization",
            "name": "ExampleNamespaceOne"
          },
          "name": "File102.txt"
        },
        "samples": {
          "kind": "Reference",
          "samples": "File102_samples"
        },
        "gateways": [
		# omitted
        ],
        "metadata": {
		# omitted
          }
        }
      }

#### File Schema
The only change would be replacing the "samples" list of sample IDs with a cohort.AnonymousOrReference instead.

models.File {
```
  id, gateways, metadata: no change
  samples: {
    models.cohort.AnonymousOrReference {}
    }
}
```

### Implementation effects
For consortium members who have a manageable number of samples per file, I would expect the interface change to be minimal.
If you aren't creating any named Cohorts, the main change is to the file endpoint, adding the additional layer in the sample list.
You would also need to add Cohort as a possible entity in your error messages, so that the Entity Not Found error message for the `/cohort/{name}` endpoint would be specific.

### Thank you!
Thanks for reading and considering all of this!

**Comment by claymcleod** [comment]
> Is a list of 10k namespace-delimited identifiers actually a large response? Is this actually a problem we need to solve?

To me, this definitely seems large enough to consider the idea.

> I have designed Cohort as a lightweight sample group. Is that what we want right now ? Is there value in going straight to a more formal grouping that reflects a scientific study or more meaningful group of samples?

I think we probably want two kinds of things:

- **Cohorts**, which are groupings of subjects.
- **Collections** (is there a better name for this?), which are groupings of samples.

> Naming - Possible names could be Cohort, Dataset, something else (sampleList?) I want the name to reflect the informality of the grouping.

See the above proposal on different groupings for subjects/samples. That's my feeling at least.

> I feel that a file should have exactly 1 cohort associated with it. Should a file have a list of cohorts instead?

I don't like this approach. It's splitting relevant concepts into too many places IMO. What I would advocate for is something like this:

```json
{
  ...,
  "samples": [{
    "kind": "Cohort",
      "namespace": {
        "organization": "example-organization",
        "name": "Namespace"
      },
      "name": "CohortName001"
  },
  {
    "kind": "Subject",
    "namespace": {
      "organization": "string",
      "name": "string"
    },
    "name": "SubjectName001"
  }],
  ...,
}
```

and

```json
{
  ...,
  "samples": [{
    "kind": "Collection",
      "namespace": {
        "organization": "example-organization",
        "name": "Namespace"
      },
      "name": "CollectionName001"
  },
  {
    "kind": "Sample",
    "namespace": {
      "organization": "string",
      "name": "string"
    },
    "name": "SampleName001"
  }],
  ...,
}

```

> Is there a compelling reason to add a /cohort endpoint that lists all cohorts known to the server? If so, how would we paginate that?

Yes, definitely. It should work as a supporting entity with all of the same functionality as those have (organizations, namespaces, etc).


> Should cohorts have a simple string for a name or should the name be qualified under organization/namespace ?

Every entity should be assigned to a namespace.

**Reply by claymcleod** [reply]
>Similar to Gateways, I do not suggest having a standalone cohort endpoint which would list cohorts known to the server. Named cohorts should instead be discoverable via the files which mention them and then be individually queryable once the name is known. In addition, a cohort list endpoint would have the same problem as the current file endpoint of potentially containing long unpaginated data.

In this case though, I think the requestor is probably willing to pay the price and expects that the output is going to be very long. So, to me, it's okay as long as we communicate that in the documentation. It's also true that no cohorts will be _duplicated_ in the `/collection` requests, whereas they would be in other entity endpoints such as `/file`.
